### ISSUE TITLE
[BUG]

### ISSUE THREAD

**Issue by AccelaMichaelHe** [Azure/azure-sdk-for-java#31977]
Hi @ibrahimrabab,

We came across similar issue but in staging multiple blocks(chunk size is about 90-100 MB per chunk), the original file could be 1 GB and user split into 10 or 10+ chunks to upload async.

When we try to do some pressure testing about sending 30 chunks requests (it is kind of 3 users upload 1 GB files at the same time), the OOM (java heap space exception) is coming up.

We are using the lib 'com.azure:azure-storage-blob:12.19.0'
Related method or code snipped:

async client:
BlockBlobAsyncClient blobAsyncClient = buildBlockBlobAsyncClient();
blobAsyncClient.stageBlock(base64BlockId, Flux.just(byteBuffer),bytes.length).block();

Or
BlockBlobClient blobClient = buildBlockBlobClient();
ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(bytes);
blockBlobClient.stageBlock(base64BlockId, byteArrayInputStream, bytes.length);

JVM setting:
-XX:+UseG1GC -Xmx2024m
JDK 18
Springboot 2.7.1
[Azure](azure-storage-blob:12.19.0)

Error:
org.springframework.web.util.NestedServletException: Handler dispatch failed; nested exception is java.lang.OutOfMemoryError: Java heap space at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1082) at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:963) at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006) at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:909) at javax.servlet.http.HttpServlet.service(HttpServlet.java:681) at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883) at javax.servlet.http.HttpServlet.service(HttpServlet.java:764) at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227) at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53) at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:204) at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:183) at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:354) at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:267) at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100) at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93) at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) at org.springframework.boot.actuate.metrics.web.servlet.WebMvcMetricsFilter.doFilterInternal(WebMvcMetricsFilter.java:96) at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201) at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:197) at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97) at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541) at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135) at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92) at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78) at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:360) at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:399) at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65) at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:890) at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1787) at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191) at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659) at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) at java.base/java.lang.Thread.run(Thread.java:833) Caused by: java.lang.OutOfMemoryError: Java heap space at java.base/java.util.Arrays.copyOf(Arrays.java:3537) at java.base/java.io.ByteArrayOutputStream.ensureCapacity(ByteArrayOutputStream.java:100) at java.base/java.io.ByteArrayOutputStream.write(ByteArrayOutputStream.java:130) at org.springframework.util.StreamUtils.copy(StreamUtils.java:167) at org.springframework.util.FileCopyUtils.copy(FileCopyUtils.java:112) at org.springframework.util.FileCopyUtils.copyToByteArray(FileCopyUtils.java:152) at org.springframework.web.multipart.support.StandardMultipartHttpServletRequest$StandardMultipartFile.getBytes(StandardMultipartHttpServletRequest.java:246) at com.accela.documentservice.service.DocumentServiceImpl.uploadChunk(DocumentServiceImpl.java:46) at com.accela.documentservice.controller.DocumentController.uploadFile(DocumentController.java:67) at com.accela.documentservice.controller.DocumentController$$FastClassBySpringCGLIB$$1e756aad.invoke(<generated>) at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218) at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:793) at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163) at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763) at org.springframework.validation.beanvalidation.MethodValidationInterceptor.invoke(MethodValidationInterceptor.java:123) at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763) at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:708) at com.accela.documentservice.controller.DocumentController$$EnhancerBySpringCGLIB$$c36abd95.uploadFile(<generated>) at java.base/java.lang.invoke.LambdaForm$DMH/0x0000000801400400.invokeVirtual(LambdaForm$DMH) at java.base/java.lang.invoke.LambdaForm$MH/0x0000000801408800.invoke(LambdaForm$MH) at java.base/java.lang.invoke.Invokers$Holder.invokeExact_MT(Invokers$Holder) at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invokeImpl(DirectMethodHandleAccessor.java:158) at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104) at java.base/java.lang.reflect.Method.invoke(Method.java:577) at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:205) at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:150) at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:117) at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:895) at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:808) at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87) at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1067)

Please check if this is a defect or any advice to use this function correctly?
Regards
Michael

**Comment by joshfree** [comment]
Thank you for filing this github issue, @AccelaMichaelHe - folks from the Storage and Azure SDK Performance group will follow up shortly.

/cc @jaschrep-msft @g2vinay @mikeharder

**Comment by AccelaMichaelHe** [comment]
Thanks! And one more issue, if we set jvm larger with just 3GB files upload , it seems ok but the memory is getting much more higher and never release. 
Looking forward to hearing your feedback. 
Regards

**Comment by jaschrep-msft** [comment]
Hi @AccelaMichaelHe, it sounds like you are following up from a previous issue you've filed, but I don't have the context of that previous issue and so am unsure if there is more I'm supposed to know when reading this.

My initial thoughts are that claim to be uploading in parallel 30 requests of ~100MB each, for a total of ~3GB you have buffered yourself. You are only setting your JVM memory to 2GiB. This seems like it would have an OOM exception no matter what. If I'm misunderstanding, please let me know.

**Comment by AccelaMichaelHe** [comment]
> Hi @AccelaMichaelHe, it sounds like you are following up from a previous issue you've filed, but I don't have the context of that previous issue and so am unsure if there is more I'm supposed to know when reading this.
> 
> My initial thoughts are that claim to be uploading in parallel 30 requests of ~100MB each, for a total of ~3GB you have buffered yourself. You are only setting your JVM memory to 2GiB. This seems like it would have an OOM exception no matter what. If I'm misunderstanding, please let me know.

Yes @jaschrep-msft , you are correct. This is the first issue(as we supposed to be set the JVM heap size might be 2GB only inside the container). And we tried to enlarge to see and we found the memory is not released only when you are trying to release it manually (Springboot Application). Please let me know if you need more information

**Comment by jaschrep-msft** [comment]
@AccelaMichaelHe Unless I'm mistaken, you are the one allocating this memory, not the Storage library. You are the one buffering several GB of byte arrays at once in your own code, judging from the snippets you've posted in the issue. Please provide the information that has led you to believe this is an issue with the Storage library.

**Comment by AccelaMichaelHe** [comment]
Below are the code snippets:
```
  public void stageBlock(UUID uploadSessionId, String base64BlockId, byte[] bytes) {
    BlockBlobClient blockBlobClient = buildBlockBlobClient(uploadSessionId, filename);
    try(ByteArrayInputStream bis = new ByteArrayInputStream(bytes)) {
      blockBlobClient.stageBlock(base64BlockId, bis, bytes.length);
    } catch (Exception e) {
      throw new StorageBlockStagingException(e);
    }
  }
```
If we do not call  **_blockBlobClient.stageBlock(base64BlockId, bis, bytes.length);_**, the memory will not getting extreamly high. But with calling the block, memory getting much more higher and almost occupy 90% (upload a large block, 1 user per time , run 10 times)and the memory will keep running for a long time.

We are writing a simple code to test only and u can try on your side for uploading the file size of 90 - 100 MB per block(chunk).

Regards,
Michael

**Comment by jaschrep-msft** [comment]
@AccelaMichaelHe I do see at least one superfluous conversion of data that can probably be optimized away. I can look further into this.

I do still maintain though that you claim to be calling this snippet 30 times in parallel with ~100MB byte arrays being passed in each time. You would have to buffer ~3GB in byte arrays to do this unless you happen to be uploading the same byte array 30 times. You would hit OOM in your own buffering code. You say this doesn't happen if you don't call stageblock, but it makes no sense that you can buffer ~3GB in byte arrays when your JVM heap size is 2GB. Something is not as you are describing it, and so it is hard for me to reason out culprit code as a result.

**Comment by AccelaMichaelHe** [comment]
> 

@jaschrep-msft Thanks for the reply, you can forget that part as you said correctly that it impossible to send the size beyond the JVM heap size. 
For the second issue as I mentioned in previous comment if you can help look into it and guide us how could we release the big object in the memory that will be helpful.
Very appreciate and looking forward your feedback.

Regards,
Michael

**Comment by jaschrep-msft** [comment]
@AccelaMichaelHe I'm not an expert on JVM garbage collection but my understanding is that the JVM will attempt to garbage collect to the highest extent it can before running into an OOM exception. If it can't release memory, then the memory is still in reference somewhere. Since clients only hold onto given memory for the life of the transfer operation, they aren't stored anywhere in the clients, themselves, then I imagine you are the one holding onto this memory. That's not something I can help you fix. You need to investigate your own memory usage. If you find evidence to suggest storage is leaking memory, feel free to open another issue with that evidence.
